bigotes.1
.Dd Apr 18, 2024
.Dt BIGOTES 1
.Os
.Sh NAME
.Nm bigotes
.Nd benchmark measurement tool
.Sh SYNOPSIS
.Nm bigotes
.Op Ar options
.Op Fl -
.Ar command
.Op Ar args...
.Nm bigotes
.Op Ar options
.Fl i
.Sh DESCRIPTION
The
.Nm
program is a versatile tool designed to perform automatic benchmark
measurements and report statistical information. A single quantity is
sampled multiple times by either running again a given program or by
reading the values from the standard input. The typical quantity 
is the execution time, but others can be used too.
.Pp
The execution of the program is repeated until at least 30 samples are
collected and 30 seconds have elapsed. Both criteria must be met to stop
collecting samples.
.Pp
At the end, a summary is presented with several statistics. An histogram
is also plotted by using UTF-8 block characters and the results of the
Shapiro-Wilk test for normality:
.Bd -literal -offset xx
% bigotes -w -s 'sleep 0.1'

       MIN         Q1     MEDIAN       MEAN         Q3        MAX
 1.022e-01  1.067e-01  1.070e-01  1.080e-01  1.106e-01  1.111e-01
         N       WALL        MAD      STDEV       SKEW   KURTOSIS
       278       30.0  1.552e-03  2.425e-03 -3.557e-01 -6.176e-01

    Run 278 times for 30.0 s: sleep 0.1
    Shapiro-Wilk: W=8.23e-01, p-value=3.97e-17 (NOT normal)

                                 █▁                          ▁
                                 ██                          █▆
                                 ██▄                        ▃██
 ▂▁▁▂ ▁  ▁▁ ▁   ▁▁ ▁     ▁    ▁▅▅███▃  ▁              ▁    ▃███▅▁

.Ed
The list of summary values is described as as follows:
.Bl -tag -width Ds
.It Em MIN
The minimum value of the sample.
.It Em Q1
The first quartile (25 percentile) of the sample.
.It Em MEDIAN
The median of the sample.
.It Em MEAN
The mean of the sample.
.It Em Q3
The third quartile (75 percentile) of the sample.
.It Em MAX
The maximum value of the sample.
.It Em N
The number of samples.
.It Em WALL
The total amount of wall clock time the program spent running.
.It Em FAR
The number of outlier values that are "far out" as defined by John
Tukey with k=3. That is, values smaller than Q1-3*(Q3-Q1) or bigger
than Q3+3*(Q3-Q1).
.It Em MAD
The median absolute deviation of the sample, scaled by 1.4826 to match
the value of the standard deviation in a normal distribution.
.It Em STDEV
The standard deviation of the sample.
.It Em SKEW
The sample skewness, a measure of the asymmetry of the probability
distribution about its mean.
.It Em KURTOSIS
The sample excess kurtosis (kurtosis minus 3). A value larger than 0
indicates a fatter tail in the probability distribution when compared to
a normal distribution, and often occurs in outliers.
.El
.Pp
.Pp
There are two modes of operation and one must be selected.
.Bl -tag -width Ds
.It Fl -
Selects the
.Em exec
mode (the default). Runs the command and arguments as-is, without a shell. The 
.Fl -
flag may be omitted.
.It Fl i
Selects the
.Em stdin
mode. Reads quantities values from the standard input, one per line.
Nothing is executed in this mode.
.El
.Pp
The
.Ar options
must precede the operation mode and are as follows:
.Bl -tag -width Ds
.It Fl o Ar file
Write the measured quantities in the given
.Ar file ,
one per line. Not compatible with the stdin mode.
.It Fl m
Format the output to be read by a machine.
.It Fl q
Be quiet, don't report progress information during execution.
.It Fl w
Instead of reading the quantity from the standard output, use the
wall clock time (in seconds) that the program took to run. Not
compatible with the stdin mode.
.It Fl X
Remove outliers from the stopping criteria and histogram. The summary
statistics always use all samples.
.It Fl t
Specify the minimum wall time in seconds (default 30).
.El
.Pp
The name
.Nm
comes from the Spanish word for whiskers which are commonly used in
statistics to represent the dispersion of a sample when drawing box
plots or error bars.
.Sh EXIT STATUS 
.Ex -std
.Sh EXAMPLES
.Bl -enum -compact -width w
.It
Measure the wall clock time taken by a command:
.Bd -literal -offset xx
% bigotes -w find ~/pic -iname '*.jpg'

       MIN         Q1     MEDIAN       MEAN         Q3        MAX
 2.769e-02  2.823e-02  2.840e-02  2.856e-02  2.867e-02  4.067e-02
         N       WALL        MAD      STDEV       SKEW   KURTOSIS
      1051       30.0  3.080e-04  6.712e-04  7.561e+00  1.110e+02

    Run 1051 times for 30.0 s: find
    Shapiro-Wilk: W=5.90e-01, p-value=3.34e-44 (NOT normal)

   ▇█
   ██▂
  ▄███▁
 ▁█████▃▄▄▃▁▁▁▁▁ ▁▁    ▁        ▁ ▁                             ▁

.Ed
.It
Run a benchmark in a remote machine by using srun and use the value
written in the standard output as measured quantity:
.Bd -literal -offset xx
% srun -u -c56 bigotes b6_heat_nanos6 -s 2048 -t 10 -b 64

       MIN         Q1     MEDIAN       MEAN         Q3        MAX
 3.098e-02  3.321e-02  3.892e-02  3.701e-02  3.962e-02  4.251e-02
         N        FAR        MAD      STDEV       SKEW   KURTOSIS
       184          0  2.984e-03  3.508e-03 -3.635e-01 -1.515e+00

    Run 184 times for 30.1 s: b6_heat_nanos6
    Shapiro-Wilk: W=8.57e-01, p-value=3.84e-12 (NOT normal)

                                              ██
                                             ▇██ ▁
   ▂  ▂▄▁▂ ▁▂▁ ▁  ▂                       ▄ ▂███▆█▁   ▁▂   ▆
 ▂▄█▆▄████▆███▆█▄▆█▂▂▄ ▂▂     ▂     ▂  ▂▂▂█▆███████▄▄▄██▆▆ █▄▂ ▂▂

.Ed
.It
Ping a remote host and take the time as the measurement using a pipe
(note the escaping for the shell):
.Bd -literal -offset xx
% bigotes sh -c "ping -c 1 1.1.1.1 | awk -F/ 'END {print \\$5}'"

       MIN         Q1     MEDIAN       MEAN         Q3        MAX
 1.489e+01  1.864e+01  1.922e+01  1.938e+01  1.969e+01  4.145e+01
         N       WALL        MAD      STDEV       SKEW   KURTOSIS
       963       30.0  7.769e-01  1.902e+00  6.103e+00  5.081e+01

    Run 963 times for 30.0 s: ping -c 1 1.1.1.1 | awk -F/ 'END {print $5}'
    Shapiro-Wilk: W=5.07e-01, p-value=1.74e-45 (NOT normal)

          ▄█▅
         ▂███
       ▁▂████▅▁
 ▁   ▁▁████████▁▁▁▁▁▁   ▁▁▁▁▁▁▁▁    ▁▁       ▁▁▁▁   ▁   ▁       ▁

.Ed
.It
Read the measurements from the standard input, generated by a Gaussian
distribution using the gsl-randist tool:
.Bd -literal -offset xx
% gsl-randist 0 5000 gaussian 0.5 | bigotes -i

       MIN         Q1     MEDIAN       MEAN         Q3        MAX
-1.736e+00 -3.227e-01  4.732e-03  1.135e-02  3.481e-01  1.669e+00
         N       WALL        MAD      STDEV       SKEW   KURTOSIS
      5000        0.0  4.986e-01  4.985e-01 -5.619e-03 -5.210e-02

    Read 5000 samples from stdin
    Shapiro-Wilk: W=1.00e+00, p-value=7.00e-01 (may be normal)

                           ▂  ▆▆█▅▅▇▂▁▁▂▁
                        ▁▄▂█▇▇███████████▅ ▁
                  ▁▁▁▆▆███████████████████▇█▇▄▃▃▁▁
 ▁ ▁▁▁▁▁▁▁▁▂▂▃▂▄▄▆████████████████████████████████▆▅▃▄▂▁▁▁▁▁▁▁▁▁▁

.Ed

Notice how the STDEV and MAD values are close.

.Sh SEE ALSO
.Pp
The repository 
.Lk https://github.com/rodarima/bigotes/
contains the last version of the program.
.Sh AUTHORS
.An "Rodrigo Arias Mallo" Aq Mt "rodrigo.arias@bsc.es"


bigotes.c
/* Copyright (c) 2021-2024 Barcelona Supercomputing Center (BSC)
 * SPDX-License-Identifier: GPL-3.0-or-later */

#include <errno.h>
#include <limits.h>
#include <math.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <time.h>
#include <unistd.h>

#include "swilk.h"
#include "stats.h"
#include "common.h"

static int read_from_stdin = 0;
static int use_wall_clock = 0;
static double min_wall_time = 30.0;
static int use_exec = 0;
static int use_machine_output = 0;
static int be_quiet = 0;
static int trim_outliers = 0;
static long min_samples = 30;
static const char *output_fname = "bigotes.csv";
static const char *reference_fname = NULL;
static int compare_mode = 0;

struct sampling {
	long nmax;
	long nmin;
	long n;
	double *samples;
	double last;
	double wall;
	char *const *cmd;
	char *const *argv;
	int argc;
	double t0;
	double min_time;
	double last_stats;
};

/* Returns the current time in seconds since some point in the past */
static double
get_time(void)
{
	struct timespec tv;
	if(clock_gettime(CLOCK_MONOTONIC, &tv) != 0)
	{
		perror("clock_gettime failed");
		exit(EXIT_FAILURE);
	}

	return (double)(tv.tv_sec) +
		(double)tv.tv_nsec * 1.0e-9;
}

static void
drain(FILE *f)
{
	char line[4096];
	while (fgets(line, 4096, f) != NULL)
		;
}

static int
process_stdout(FILE *f, double *metric)
{
	if (use_wall_clock) {
		drain(f);
		return 0;
	}

	char line[4096];
	if (fgets(line, 4096, f) == NULL) {
		err("missing stdout line");
		return -1;
	}

	char *nl = strchr(line, '\n');
	if (nl != NULL)
		*nl = '\0';

	if (sscanf(line, "%le", metric) != 1) {
		err("failed to read metric from stdout");
		return -1;
	}

	/* Drain the rest of the stdout */
	while (fgets(line, 4096, f) != NULL)
		;

	return 0;
}

static int
do_exec(char *const argv[], double *metric)
{
	int fd[2];
	const int R = 0, W = 1;

	if (pipe(fd) == -1) {
		err("pipe failed:");
		return -1;
	}

	pid_t pid = fork();
	if (pid < 0) {
		err("fork failed:");
		return -1;
	} else if (pid > 0) {
		/* Parent */
		close(fd[W]);
		FILE *f = fdopen(fd[R], "r");
		if (f == NULL) {
			err("fdopen failed:");
			return -1;
		}
		if (process_stdout(f, metric) != 0)
			return -1;
		fclose(f);
	} else {
		/* Child */
		/* Make the stdout go to the pipe */
		if (dup2(fd[W], STDOUT_FILENO) == -1) {
			err("dup2 failed:");
			return -1;
		}
		close(fd[W]);
		close(fd[R]);

		if (execvp(argv[0], argv) != 0) {
			err("execvp failed:");
			return -1;
		}

		/* Not reached */
		err("execvp unexpected return");
		return -1;
	}

	return 0;
}

static int
cmp_double(const void *pa, const void *pb)
{
	double a = *(const double *) pa;
	double b = *(const double *) pb;

	if (a < b)
		return -1;
	else if (a > b)
		return 1;
	else
		return 0;
}

/* Test for normality */
static void
shapiro_wilk_test(struct sampling *s)
{
	double W, p;
	if (swilk(s->samples, s->n, &W, &p) != 0) {
		err("swilk failed");
		return;
	}

	if (use_machine_output) {
		printf("%-10s %e\n", "sw_pvalue", p);
		printf("%-10s %d\n", "sw_normal", (p >= 0.05));
	} else {
		const char *msg = (p < 0.05) ? "NOT normal" : "may be normal";
		printf("    Shapiro-Wilk: W=%.2e, p-value=%.2e (%s)\n", W, p, msg);
	}

}

static void
stats(struct sampling *s)
{
	if (s->n < 1)
		return;

	long outliers = 0;
	double last = s->samples[s->n - 1];
	double median = last;
	double mean = last;
	double var = NAN;
	double stdev = NAN;
	double sem = NAN;
	double rsem = NAN;
	double mad = NAN;
	double q1 = NAN;
	double q3 = NAN;
	double iqr = NAN;

	/* Need at least two samples */
	if (s->n >= 2) {
		/* Sort samples to take the median */
		qsort(s->samples, s->n, sizeof(double), cmp_double);

		double *absdev = safe_calloc(s->n, sizeof(double));

		q1     = s->samples[s->n / 4];
		median = s->samples[s->n / 2];
		q3     = s->samples[(s->n * 3) / 4];

		iqr = q3 - q1;

		double sum = 0.0;
		long ncorr = 0;
		for (long i = 0; i < s->n; i++) {
			double x = s->samples[i];
			if (trim_outliers) {
				if (x < q1 - 3.0 * iqr || x > q3 + iqr * 3.0)
					continue;
			}
			sum += x;
			ncorr++;
		}

		mean = sum / (double) ncorr;
		double sumsqr = 0.0;
		for (long i = 0; i < s->n; i++) {
			double x = s->samples[i];
			double dev = x - mean;
			absdev[i] = fabs(s->samples[i] - median);
			//printf("absdev[%3ld] = %e\n", i, absdev[i]);
			if (x < q1 - 3.0 * iqr || x > q3 + iqr * 3.0) {
				outliers++;
				if (trim_outliers)
					continue;
			}
			sumsqr += dev * dev;
		}
		qsort(absdev, s->n, sizeof(double), cmp_double);
		mad = absdev[s->n / 2];

		var = sumsqr / ncorr;
		stdev = sqrt(var);
		sem = stdev / sqrt(ncorr);
		rsem = 100.0 * sem * 1.96 / mean;
		free(absdev);
	}

	double rmad = 100.0 * mad / median;

	if (!be_quiet) {
		fprintf(stderr, "\rn=%ld t=%.1fs median=%.2e rmad=%.2f%% rsem=%.2f%% far=%ld    ",
				s->n, s->wall, median, rmad, rsem, outliers);
		fflush(stderr);
	}

	s->last_stats = get_time();
}

static int
should_continue(struct sampling *s)
{
	double dt = get_time() - s->last_stats;
	/* Update stats at 10 FPS */
	if (dt > 1.0 / 10.0)
		stats(s);

	if (s->n >= s->nmax)
		return 0;

	if (s->n < s->nmin)
		return 1;

	if (s->wall < s->min_time)
		return 1;

	return 0;
}

static void
load_sample(struct sampling *s, double metric, double walltime)
{
	if (s->n >= s->nmax) {
		err("too many samples");
		exit(1);
	}

	s->samples[s->n] = metric;
	s->n++;
	s->last = metric;
	s->wall += walltime;
}

static void
write_sample(double metric)
{
	if (!read_from_stdin) {
		FILE *f = fopen(output_fname, "a");
		if (f == NULL) {
			err("fopen failed:");
			exit(1);
		}
		fprintf(f, "%e\n", metric);
		fclose(f);
	}
}

static void
hist_bounds(struct sampling *s, int cut_outliers, double *pmin, double *pmax)
{
	qsort(s->samples, s->n, sizeof(double), cmp_double);

	double qmin = 0.0, qmax = 0.0;
	if (cut_outliers) {
		double q1 = s->samples[s->n / 4];
		double q3 = s->samples[(s->n * 3) / 4];
		double iqr = q3 - q1;
		qmin = q1 - 3.0 * iqr;
		qmax = q3 + 3.0 * iqr;
	} else {
		qmin = s->samples[0];
		qmax = s->samples[s->n - 1];
	}

	dbg("qmin=%e, qmax=%e", qmin, qmax);

	/* Expand limits if needed */
	if (isnan(*pmin) || qmin < *pmin) {
		dbg("setting *pmin=qmin=%e", qmin);
		*pmin = qmin;
	}

	if (isnan(*pmax) || qmax > *pmax) {
		dbg("setting *pmax=qmax=%e", qmax);
		*pmax = qmax;
	}

	dbg("*pmin=%e, *pmax=%e", *pmin, *pmax);
}

static void
compute_histogram(struct sampling *s, int nbins, long *count, double qmin, double qmax)
{
	qsort(s->samples, s->n, sizeof(double), cmp_double);

	double binlen = (qmax - qmin) / (double) nbins;

	long j = 0;

	for (int i = 0; i < nbins; i++) {
		double binstart = qmin + (double) (i + 0) * binlen;
		double binend   = qmin + (double) (i + 1) * binlen;
		count[i] = 0;

		for (; j < s->n; j++) {
			double x = s->samples[j];
			if (x < binstart) {
				/* Left out */
				continue;
			}

			if (x > binend) {
				/* Switch bin */
				break;
			}

			/* Fits this bin */
			count[i]++;
		}
	}
}

static void
draw_cell(double q)
{
	char *utf8[] = { "▁", "▂", "▃", "▄", "▅", "▆", "▇", "█"};

	int i = (q * 8.0) - 1;
	if (i > 7)
		i = 7;
	else if (i < 0)
		i = 0;

	printf("%s", utf8[i]);
}

static void
plot_histogram(struct sampling *s, int w, int h, int cut_outliers, double qmin, double qmax)
{
	long *count = safe_calloc(w, sizeof(long));

	compute_histogram(s, w, count, qmin, qmax);

	long maxcount = 0;
	for (int i = 0; i < w; i++) {
		if (count[i] > maxcount)
			maxcount = count[i];
	}

	double *barlen = safe_calloc(w, sizeof(double));
	for (int i = 0; i < w; i++) {
		double rel = (double) count[i] / (double) maxcount;
		barlen[i] = (double) h * rel;
	}

	for (int i = h-1; i >= 0; i--) {
		if (i == 0 && cut_outliers) {
			printf("…");
		} else {
			putchar(' ');
		}
		for (int j = 0; j < w; j++) {
			double cell = barlen[j] - i;
			if (cell > 0.0)
				draw_cell(cell);
			else
				putchar(' ');
		}
		if (i == 0 && cut_outliers) {
			printf("…");
		}
		putchar('\n');
	}

	free(barlen);
	free(count);
}

static void
print_summary(struct sampling *s)
{
	qsort(s->samples, s->n, sizeof(double), cmp_double);

	double mean = stats_mean(s->samples, s->n);
	double stdev = stats_stdev(s->samples, s->n, mean);
	double skewness = stats_skewness(s->samples, s->n, mean, stdev);
	double kurtosis = stats_kurtosis(s->samples, s->n, mean, stdev);

	double xmin = stats_percentile(s->samples, s->n, 0.0);
	double q1 = stats_percentile(s->samples, s->n, 0.25);
	double median = stats_median(s->samples, s->n);
	double q3 = stats_percentile(s->samples, s->n, 0.75);
	double xmax = stats_percentile(s->samples, s->n, 1.0);

	double mad = stats_mad(s->samples, s->n, median);
	long n = s->n;
	long far = stats_outliers(s->samples, s->n, q1, q3, 3.0);

	double sem = stats_sem(stdev, s->n);
	double rel_stdev = stats_percent(stdev, mean);
	double rel_mad = stats_percent(mad, median);
	double rel_sem = stats_percent(sem * 1.96, mean);
	double rel_far = stats_percent(far, s->n);

	if (use_machine_output) {
		printf("%-10s %e\n", "min", xmin);
		printf("%-10s %e\n", "q1", q1);
		printf("%-10s %e\n", "median", median);
		printf("%-10s %e\n", "mean", mean);
		printf("%-10s %e\n", "q3", q3);
		printf("%-10s %e\n", "max", xmax);
		printf("%-10s %ld\n", "samples", n);
		printf("%-10s %e\n", "wall", s->wall);
		printf("%-10s %e\n", "mad", mad);
		printf("%-10s %e\n", "stdev", stdev);
		printf("%-10s %e\n", "skewness", skewness);
		printf("%-10s %e\n", "kurtosis", kurtosis);
		printf("%-10s %ld\n", "far", far);
		printf("%-10s %e\n", "rfar", rel_far);
		printf("%-10s %e\n", "rmad", rel_mad);
		printf("%-10s %e\n", "rstdev", rel_stdev);
		printf("%-10s %e\n", "sem", sem);
		printf("%-10s %e\n", "rsem", rel_sem);
	} else {
		printf("\n");
		printf("%10s %10s %10s %10s %10s %10s\n",
				"MIN", "Q1", "MEDIAN", "MEAN", "Q3", "MAX");
		printf("% 10.3e % 10.3e % 10.3e % 10.3e % 10.3e % 10.3e \n",
				xmin, q1, median, mean, q3, xmax);
		printf("\n");
		printf("%10s %10s %10s %10s %10s %10s\n",
				"N", "WALL", "MAD", "STDEV", "SKEW", "KURTOSIS");
		printf("%10ld %10.1f % 10.3e % 10.3e % 10.3e % 10.3e\n",
				n, s->wall, mad, stdev, skewness, kurtosis);
		printf("\n");
		printf("%10s %10s %10s %10s %10s %10s\n",
				"FAR", "%FAR", "%MAD", "%STDEV", "SEM", "%SEM");
		printf("%10ld %10.2f % 10.2f % 10.2f % 10.3e % 10.2f\n",
				far, rel_far, rel_mad, rel_stdev, sem, rel_sem);
		printf("\n");
	}
}


/* Return -1 on error, 0 on success */
static int
do_read(FILE *f, double *metric, int *end)
{
	char line[4096];
	if (fgets(line, 4096, f) == NULL) {
		if (feof(f)) {
			*end = 1;
			return 0;
		}

		err("missing stdout line");
		return -1;
	}

	char *nl = strchr(line, '\n');
	if (nl != NULL)
		*nl = '\0';

	if (sscanf(line, "%le", metric) != 1) {
		err("cannot find number in: %s", line);
		return -1;
	}

	return 0;
}

static void
print_command(struct sampling *s)
{
	if (use_machine_output) {
		const char *runmode = read_from_stdin ? "stdin" : "exec";
		printf("%-10s %s\n", "runmode", runmode);

		if (!read_from_stdin) {
			printf("%-10s ", "command");
			for (char *const* p = s->cmd; *p; p++) {
				printf("%s ", *p);
			}
			printf("\n");
		}
		return;
	}

	if (read_from_stdin) {
		printf("    Read from stdin\n");
	} else {
		printf("    Cmd: ");
		for (char *const* p = s->cmd; *p; p++) {
			printf("%s ", *p);
		}
		printf("\n");
	}
}

static int
do_sample(char * const cmd[], char * const argv[], int argc)
{
	struct sampling ref = { 0 };

	ref.nmax = 5000;
	ref.samples = safe_calloc(ref.nmax, sizeof(double));

	/* First read reference if in compare mode */
	if (compare_mode) {
		FILE *f = fopen(reference_fname, "r");
		if (f == NULL) {
			err("fopen failed:");
			exit(1);
		}
		while (1) {
			double metric;
			int end = 0;
			if (do_read(f, &metric, &end) != 0) {
				err("cannot read sample from %s:",
						reference_fname);
				return 1;
			}

			if (end)
				break;

			load_sample(&ref, metric, 0.0);
		}
		fclose(f);
	}

	struct sampling s = { 0 };
	s.nmax = 5000;
	s.nmin = min_samples;
	s.min_time = min_wall_time;
	s.samples = safe_calloc(s.nmax, sizeof(double));
	s.n = 0;
	s.cmd = cmd;
	s.argv = argv;
	s.argc = argc;
	s.t0 = get_time();

	if (!read_from_stdin) {
		FILE *f = fopen(output_fname, "w");
		if (f == NULL) {
			err("fopen failed:");
			exit(1);
		}
		fclose(f);
	}

	while (read_from_stdin || should_continue(&s)) {
		double metric;
		double walltime = 0.0;

		if (read_from_stdin) {
			int end = 0;
			if (do_read(stdin, &metric, &end) != 0) {
				err("cannot read sample from stdin");
				return 1;
			}

			if (end)
				break;
		} else {
			double t0 = get_time();
			if (do_exec(cmd, &metric) != 0)
				return 1;
			double t1 = get_time();
			walltime = t1 - t0;
			if (use_wall_clock)
				metric = walltime;
		}

		load_sample(&s, metric, walltime);
		write_sample(metric);
	}

	/* Always recompute the stats with all samples */
	if (!read_from_stdin && !be_quiet) {
		/* Clear stat line */
		fprintf(stderr, "\r                                                            \r");
		fflush(stderr);
	}

	print_summary(&s);
	print_command(&s);

	shapiro_wilk_test(&s);

	double qmin = NAN;
	double qmax = NAN;

	printf("\n"); /* Leave one empty before histogram */

	if (compare_mode) {
		dbg("s->n=%ld ref->n=%ld", s.n, ref.n);

		hist_bounds(&s, trim_outliers, &qmin, &qmax);
		dbg("bounds: %e %e", qmin, qmax);
		hist_bounds(&ref, trim_outliers, &qmin, &qmax);
		dbg("bounds: %e %e", qmin, qmax);
		plot_histogram(&s, 64, 4, trim_outliers, qmin, qmax);
		printf("\n");
		printf("Ref: %s\n", reference_fname);
		printf("\n");
		plot_histogram(&ref, 64, 4, trim_outliers, qmin, qmax);
	} else {
		hist_bounds(&s, trim_outliers, &qmin, &qmax);
		plot_histogram(&s, 64, 4, trim_outliers, qmin, qmax);
	}

	printf("\n");

	free(s.samples);

	return 0;
}

static void
usage(void)
{
	fprintf(stderr, "Usage:\n");
	fprintf(stderr, "    %s [options] [--] COMMAND [ARGS...]\n", progname_get());
	fprintf(stderr, "    %s [options] -i\n", progname_get());
	fprintf(stderr, "See the %s(1) manual for more details.\n", progname_get());
	exit(1);
}

int
main(int argc, char *argv[])
{
	progname_set("bigotes");
	int opt;

	while ((opt = getopt(argc, argv, "imn:wo:qhXt:c:")) != -1) {
		switch (opt) {
			case 'i':
				read_from_stdin = 1;
				break;
			case 'w':
				use_wall_clock = 1;
				break;
			case 'o':
				output_fname = optarg;
				break;
			case 'c':
				reference_fname = optarg;
				compare_mode = 1;
				break;
			case 'q':
				be_quiet = 1;
				break;
			case 'm':
				use_machine_output = 1;
				break;
			case 'n':
				min_samples = atol(optarg);
				break;
			case 't':
				min_wall_time = atof(optarg);
				break;
			case 'X':
				trim_outliers = 1;
				break;
			default: /* '?' */
				err("unknown option '%c'", opt);
				/* fall-through */
			case 'h':
				usage();
		}
	}

	if (use_exec + read_from_stdin > 1) {
		err("bad usage: only one operation mode allowed");
		usage();
	}

	if (!use_exec && !read_from_stdin)
		use_exec = 1;

	char * const * cmd = NULL;
	if (!read_from_stdin) {
		if (optind >= argc) {
			err("missing command to run");
			usage();
		} else {
			cmd = &argv[optind];
		}
	}

	if (do_sample(cmd, argv, argc) != 0)
		return 1;

	return 0;
}
